Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift

```diff
@@ -59,15 +59,15 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
 
     private let backgroundLogger: Logger
     private var logger: Logger
+    private let eventLoop: EventLoop
+    private let connectionIdLoggerMetadata: Logger.MetadataValue
 
-    let connection: HTTP1Connection
-    let eventLoop: EventLoop
-
-    init(connection: HTTP1Connection, eventLoop: EventLoop, logger: Logger) {
-        self.connection = connection
+    var onRequestCompleted: () -> Void = {}
```

Comment:
Why are we adding this indirection here?

Reply:
We need it to break the dependency between `HTTP1Connection` and `HTTP1ClientChannelHandler`. `HTTP1ClientChannelHandler` only needs a reference to `HTTP1Connection` to get the connection id for logging and to signal that the last request is complete. The `connection.id` is now passed as `connectionIdLoggerMetadata: Logger.MetadataValue` in the init. `connection.taskComplete()` is replaced with `onRequestCompleted()` indirection.

I think in the long run we should replace this with a write promise we pass together with the request here:
https://github.com/swift-server/async-http-client/blob/817d9aa99fe4c71e7bea18a9a74bc7a5fc166692/Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1Connection.swift#L101
This will cost us an additional allocation per request but this should be fine.

Reply:
Yes, this was my reaction: having unstructured callbacks like this is a bit of a worry for me. Promises are a substantial improvement.

Reply:
Is there any reason not to make that change soon?

Reply:
Ok, I'm asking only because I suspect this mystery-meat callback will live here forever if we don't make the change now.

Reply:
So I have tried to do that. The problem I run into is that it is currently not really used like a promise and may never be called. It is used to signal that the connection is again able to receive further request which it may never be for various reasons. This case is currently handled by closing the channel instead of calling this callback. At the end it just calls down to `connection.delegate.http1ConnectionReleased(connection)`

I think changing and testing this is quite some effort which I would like not to do right now.

Reply:
I have at least renamed it to `onConnectionIdle` to better explain what it actually does: https://github.com/swift-server/async-http-client/pull/658/commits/5f5dc1e3e5c2d0610bf0a8a18b83d04034cdd8b3